base/dict.jl

```diff
@@ -713,7 +713,7 @@ function skip_deleted(h::Dict, i)
 end
 
 start(t::Dict) = skip_deleted(t, 1)
-done(t::Dict, i) = done(t.vals, i)
+done(t::Dict, i) = i > length(t.vals)
```

Comment:
This changes the iterator state to be a tuple of `(IndexIterator, IteratorState)` instead of just the next index.  Dict was (incorrectly) using a state that did not come from `start`.